tools/hlsl_generator/gen.py

```diff
@@ -0,0 +1,256 @@
+                        case "'Pointer'":
+                            if options.shape == Shape.PTR_TEMPLATE:
+                                args.append("P " + op_name)
+                            elif options.shape == Shape.PSB_RT:
+                                args.append("pointer_t<spv::StorageClassPhysicalStorageBuffer, T> " + op_name)
+                            else:
+                                args.append("[[vk::ext_reference]] T " + op_name)
+                        case "'Value'" | "'Object'" | "'Comparator'" | "'Base'" | "'Insert'":
+                            args.append("T " + op_name)
+                        case "'Offset'" | "'Count'" | "'Id'" | "'Index'" | "'Mask'" | "'Delta'":
+                            args.append("uint32_t " + op_name)
+                        case "'Predicate'": args.append("bool " + op_name)
+                        case "'ClusterSize'":
+                            if "quantifier" in operand and operand["quantifier"] == "?": continue # TODO: overload
+                            else: return # TODO
+                        case _: return # TODO
+                case "IdScope": args.append("uint32_t " + op_name.lower() + "Scope")
+                case "IdMemorySemantics": args.append(" uint32_t " + op_name)
+                case "GroupOperation": args.append("[[vk::ext_literal]] uint32_t " + op_name)
+                case "MemoryAccess":
+                    writeInst(writer, templates, name, conds, result_ty, args + ["[[vk::ext_literal]] uint32_t memoryAccess"])
+                    writeInst(writer, templates, name, conds, result_ty, args + ["[[vk::ext_literal]] uint32_t memoryAccess, [[vk::ext_literal]] uint32_t memoryAccessParam"])
+                    writeInst(writer, templates + ["uint32_t alignment"], name, conds, result_ty, args + ["[[vk::ext_literal]] uint32_t __aligned = /*Aligned*/0x00000002", "[[vk::ext_literal]] uint32_t __alignment = alignment"])
+                case _: return # TODO
+
+    writeInst(writer, templates, name, conds, result_ty, args)
+
+
+def writeInst(writer: io.TextIOWrapper, templates, name, conds, result_ty, args):
+    fn_name = name[2].lower() + name[3:]
+    writer.write("template<" + ", ".join(templates) + ">\n[[vk::ext_instruction(spv::" + name + ")]]\n")
+    if len(conds) > 0:
+        writer.write("enable_if_t<" + " && ".join(conds) + ", " + result_ty + ">")
+    else:
+        writer.write(result_ty)
+    writer.write(" " + fn_name + "(" + ", ".join(args) + ");\n\n")
+
+
+if __name__ == "__main__":
+    script_dir_path = os.path.abspath(os.path.dirname(__file__))
+
+    parser = ArgumentParser(description="Generate HLSL from SPIR-V instructions")
+    parser.add_argument("output", type=str, help="HLSL output file")
+    parser.add_argument("--grammer", required=False, type=str, help="Input SPIR-V grammer JSON file", default=os.path.join(script_dir_path, "../../include/spirv/unified1/spirv.core.grammar.json"))
```

Comment:
if you want to hardcode relative path then make sure it works if I change CWD to execute the script

Reply:
ah `script_dir_path` I see - ok good